server/src/main/java/io/crate/execution/engine/JobLauncher.java

```diff
@@ -142,6 +147,18 @@
         }
     }
 
+    public static void execute(RowConsumer consumer,
+                               TransactionContext txnCtx,
+                               BiConsumer<RowConsumer, TransactionContext> biConsumer,
+                               boolean waitForCompletion) {
+        if (waitForCompletion) {
+            biConsumer.accept(consumer, txnCtx);
+        } else {
+            biConsumer.accept(new CollectingRowConsumer<>(Collectors.counting()), txnCtx);
+            consumer.accept(InMemoryBatchIterator.of(Row1.ROW_COUNT_UNKNOWN, SENTINEL), null);
+        }
+    }
```

Comment:
I think now that it is in the `JobLauncher` it would make sense to make it non-static and remove the `biConsumer` argument because it is currently always parameterised as `jobLauncher::execute`

Reply:
@mfussenegger Then the unit test could be tricky. I would need to have an instance of JobLauncher !?

Reply:
@mfussenegger I made the new `execute` non-static but kept the biconsumer as @seut suggested in early review for easy unit testing.

Reply:
That's kinda taking the disadvantages of both - you still need the `JobLauncher` instance but don't take advantage of having access to the instance variables.

I find the API/abstraction rather confusing in this state. The method is called `execute` but what it does actually depends on generic `biConsumer` parameter.

You could do something like this:

```patch
diff --git a/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java b/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java
index fdfbf954174..46af3c81cb2 100644
--- a/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java
+++ b/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java
@@ -61,13 +61,13 @@ public CompletableFuture<?> completionFuture() {
 
     public List<Object[]> getResult() throws Exception {
         return getResult(10);
     }
 
-    public List<Object[]> getResult(int timeout) throws Exception {
+    public List<Object[]> getResult(int timeout_in_ms) throws Exception {
         try {
-            return consumer.completionFuture().get(timeout, TimeUnit.SECONDS);
+            return consumer.completionFuture().get(timeout_in_ms, TimeUnit.MILLISECONDS);
         } catch (ExecutionException e) {
             Throwable cause = e.getCause();
             if (cause != null) {
                 Exceptions.rethrowUnchecked(cause);
             }
diff --git a/server/src/main/java/io/crate/execution/engine/JobLauncher.java b/server/src/main/java/io/crate/execution/engine/JobLauncher.java
index dc88413ed84..19e83483eb0 100644
--- a/server/src/main/java/io/crate/execution/engine/JobLauncher.java
+++ b/server/src/main/java/io/crate/execution/engine/JobLauncher.java
@@ -99,11 +99,11 @@
  *            N1<----------------+  and passed into a DistResultRXTask
  *             |
  *          BatchConsumer
  * </pre>
  **/
-public final class JobLauncher {
+public class JobLauncher {
 
     private final TransportJobAction transportJobAction;
     private final TransportKillJobsNodeAction transportKillJobsNodeAction;
     private final List<NodeOperationTree> nodeOperationTrees;
     private final UUID jobId;
@@ -145,18 +145,15 @@ public final class JobLauncher {
                 }
             }
         }
     }
 
-    public void execute(RowConsumer consumer,
-                        TransactionContext txnCtx,
-                        BiConsumer<RowConsumer, TransactionContext> biConsumer,
-                        boolean waitForCompletion) {
+    public void execute(RowConsumer consumer, TransactionContext txnCtx, boolean waitForCompletion) {
         if (waitForCompletion) {
-            biConsumer.accept(consumer, txnCtx);
+            execute(consumer, txnCtx);
         } else {
-            biConsumer.accept(new CollectingRowConsumer<>(Collectors.counting()), txnCtx);
+            execute(new CollectingRowConsumer<>(Collectors.counting()), txnCtx);
             consumer.accept(InMemoryBatchIterator.of(Row1.ROW_COUNT_UNKNOWN, SENTINEL), null);
         }
     }
 
     public void execute(RowConsumer consumer, TransactionContext txnCtx) {
diff --git a/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java b/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java
index d07a32ecb27..06e51d83ca1 100644
--- a/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java
+++ b/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java
@@ -20,15 +20,27 @@
  */
 
 package io.crate.execution.engine;
 
 import static org.hamcrest.core.Is.is;
+
 import java.io.IOException;
 import java.util.List;
+import java.util.UUID;
 import java.util.concurrent.TimeoutException;
+
+import org.elasticsearch.indices.IndicesService;
 import org.junit.Before;
 import org.junit.Test;
+import org.mockito.Mockito;
+
+import io.crate.data.RowConsumer;
+import io.crate.execution.jobs.JobSetup;
+import io.crate.execution.jobs.TasksService;
+import io.crate.execution.jobs.kill.TransportKillJobsNodeAction;
+import io.crate.execution.jobs.transport.TransportJobAction;
+import io.crate.metadata.TransactionContext;
 import io.crate.planner.PlannerContext;
 import io.crate.test.integration.CrateDummyClusterServiceUnitTest;
 import io.crate.testing.SQLExecutor;
 import io.crate.testing.TestingRowConsumer;
 
@@ -40,32 +52,37 @@ public class JobLauncherWaitForCompletionTest extends CrateDummyClusterServiceUn
     @Before
     public void setupExecutor() throws IOException {
         SQLExecutor e = SQLExecutor.builder(clusterService).build();
         plannerContext = e.getPlannerContext(clusterService.state());
         jobLauncher = new JobLauncher(
-            null,
-            null,
-            null,
-            null,
-            null,
-            null,
-            null,
+            UUID.randomUUID(),
+            clusterService,
+            Mockito.mock(JobSetup.class),
+            Mockito.mock(TasksService.class),
+            Mockito.mock(IndicesService.class),
+            Mockito.mock(TransportJobAction.class),
+            Mockito.mock(TransportKillJobsNodeAction.class),
             List.of(),
             false,
-            null);
+            THREAD_POOL.generic()
+        ) {
+
+            @Override
+            public void execute(RowConsumer consumer, TransactionContext txnCtx) {
+            }
+        };
     }
 
     @Test
-    public void testCopyPlanNoWaitForCompletion() throws Exception {
+    public void test_execute_without_wait_for_completion_yields_immediate_result() throws Exception {
         TestingRowConsumer consumer = new TestingRowConsumer();
-        jobLauncher.execute(consumer, plannerContext.transactionContext(), (a,b)-> {}, false);
-        assertThat((Long)consumer.getResult(1).get(0)[0], is(-1L));
-
+        jobLauncher.execute(consumer, plannerContext.transactionContext(), false);
+        assertThat((Long)consumer.getResult(50).get(0)[0], is(-1L));
     }
 
-    @Test(expected = TimeoutException.class)
-    public void testCopyPlanWaitForCompletion() throws Exception {
+    @Test
+    public void test_execute_with_wait_for_completion_timeouts() throws Exception {
         TestingRowConsumer consumer = new TestingRowConsumer();
-        jobLauncher.execute(consumer, plannerContext.transactionContext(), (a,b)-> {}, true);
-        consumer.getResult(1);
+        jobLauncher.execute(consumer, plannerContext.transactionContext(), true);
+        assertThrows(TimeoutException.class, () -> consumer.getResult(100));
     }
 }
 
 ```